### PR TITLE
new notifications type: SLA breach combined (per product)

### PR DESCRIPTION
**Pull request by pna-nca**
**Description**

This commit introduces a new type of notifications: SLA breach combined.

The main difference is that notification is produced per product. Original SLA breach notifications are generated for each applicable findings. This may result in hundreds of messages (e-mail, slack or teams messages) for large products. Such alerts are hardly manageable and in the end not of much use.

With SLA breach combined notifications being enabled a user receives a message per product with a list of findings which breach their SLA. It can be summarized in the following manner:
```
subject: <SLA breach kind> <product type> <product>

body: <product summary> <list of findings>
```

**Notes**
I understand that this pull request is not acceptable as it is.
Could you please comment if the approach and motivation is acceptable? For instance, may be it is possible to avoid SLA breach notification per finding completely and make all of them "combined"?

I can work on further improvements, however, internally, we will use this feature in its current implementation, potentially, with minor updates.

### CODE REVIEW

**Comment by kiblik** [comment]
In general, I do not see any critical part where I wouldn’t agree with implementation.

But there are failing unittests + flake8. Can you look at it, please?

Plus, I know that it is a bit tricky to write unittest for notifications. Can you at least provide some example screenshots, of how this notification email can look like?

Good job. If PR is accepted, I might add and test the template for Slack.

**Comment by pna-nca** [comment]
> But there are failing unittests + flake8. Can you look at it, please?

I'll check.

> Can you at least provide some example screenshots, of how this notification email can look like?

Sure, here is the example of a notification e-mail:
image

File: dojo/templates/notifications/mail/sla_breach_combined.tpl [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+                {% elif breach_kind == 'breaching' %}
+                  {% blocktranslate trimmed %}
+                    These security findings breaching their SLA today:
+                  {% endblocktranslate %}
+                {% else %}
+                    This should not happen, check 'breach_kind' and 'kind' properties value in the source code.
+                {% endif %}
+                {% for f in findings %}
+                <br/>
+                  - <a href="{{base_url|full_url}}finding/{{f.id}}">{{f.id}}</a>: "{{f.title}}" ({{f.severity}} {% trans "severity" %}), {% trans "SLA age" %}: {{f.sla_age}}
```
Comment (kiblik) [inline dojo/templates/notifications/mail/sla_breach_combined.tpl:41]: I haven't noticed it before, I will comment on it here:
- Writing hardcoded `finding/{{f.id}}` is not the best practice. Can you use `{% url 'view_finding' finding.id as finding_url %}` + `href="{{ finding_url|full_url }}"`?
- Is the ID of the finding necessary to show in the message? I would replace it with `<a href="...">"{{f.title}}" ({{f.severity}} {% trans "severity" %}), {% trans "SLA age" %}: {{f.sla_age}}</a>`
- It is better to use `<ul>` + `<li>` in stread of `<br>` + `-` (same for product summary)

**Comment by pna-nca** [comment]
Here is the updated notification e-mail with `<ul>` + `<li>` for lists, URL to findings without IDs, just titles:
image

I've also updated `utils.py` to make Flake test happy.

However, I don't get why `test-rest-framework` fails with `"Unknown column 'sla_breach_combined' in 'field list'"` error in `unittests.test_migrations.TestOptiEndpointStatus.test_after_migration`. The migration is applied, so it should be fine...

**Comment by kiblik** [comment]
I have no idea why `unittests.test_migrations.TestOptiEndpointStatus.test_after_migration` is breaking your tests (and it is not breaking anything else).
But if it is the only issue, after accepting this PR https://github.com/DefectDojo/django-DefectDojo/pull/8666, it should work for you (if there is not any other issue).

**Comment by pna-nca** [comment]
> I have no idea why `unittests.test_migrations.TestOptiEndpointStatus.test_after_migration` is breaking your tests (and it is not breaking anything else). But if it is the only issue, after accepting this PR #8666, it should work for you (if there is not any other issue).

Thanks. :)

NB: I've accidentally clicked on "request review", sorry. Really vague interface in this regard...

**Comment by kiblik** [comment]
@pna-nca, can you try to rebase from the new `dev`? It looks like there was some different issue.

**Comment by pna-nca** [comment]
> @pna-nca, can you try to rebase from the new `dev`? It looks like there was some different issue.

done

**Comment by mtesauro** [comment]
@pna-nca Are you able to fix the failing tests so this can be merged?

**Comment by pna-nca** [comment]
> @pna-nca Are you able to fix the failing tests so this can be merged?

Hi @mtesauro, done.

**Review by mtesauro — APPROVED** [review]
Approved

**Review by grendel513 — APPROVED** [review]

**Review by cneill — APPROVED** [review]

**Review by blakeaowens — APPROVED** [review]

**Comment by kiblik** [comment]
This PR just broke `dev` branch. There are two `0195` db migrations. Needs to be fixed ASAP.

**Comment by mtesauro** [comment]
@kiblik Understood - saw #9312 will approve as soon as tests pass

**Pull request merged**